pkg/actions/karpenter/tasks.go

```diff
@@ -0,0 +1,142 @@
+)
+
+const (
+	kubernetesTagFormat = "kubernetes.io/cluster/%s"
+)
+
+type karpenterTask struct {
+	info               string
+	stackManager       manager.StackManager
+	cfg                *api.ClusterConfig
+	ec2API             ec2iface.EC2API
+	karpenterInstaller karpenter.InstallKarpenter
+}
+
+func (k *karpenterTask) Describe() string { return k.info }
+func (k *karpenterTask) Do(errs chan error) error {
+	return k.createKarpenterTask(errs)
+}
+
+// NewTasksToInstallKarpenter defines tasks required to create Karpenter
+func NewTasksToInstallKarpenter(cfg *api.ClusterConfig, stackManager manager.StackManager, ec2API ec2iface.EC2API, karpenterInstaller karpenter.InstallKarpenter) *tasks.TaskTree {
+	taskTree := &tasks.TaskTree{Parallel: true}
+	taskTree.Append(&karpenterTask{
+		info:               fmt.Sprintf("create karpenter for stack %q", cfg.Metadata.Name),
+		stackManager:       stackManager,
+		cfg:                cfg,
+		ec2API:             ec2API,
+		karpenterInstaller: karpenterInstaller,
+	})
+	return taskTree
+}
+
+// createKarpenterTask creates Karpenter
+func (k *karpenterTask) createKarpenterTask(errs chan error) error {
+	name := k.makeKarpenterStackName()
+
+	logger.Info("building nodegroup stack %q", name)
+	stack := builder.NewKarpenterResourceSet(k.cfg)
+	if err := stack.AddAllResources(); err != nil {
+		return err
+	}
+	tags := map[string]string{
+		api.KarpenterNameTag: name,
+	}
+	if err := k.stackManager.CreateStack(name, stack, tags, nil, errs); err != nil {
+		return fmt.Errorf("failed to create stack: %w", err)
+	}
+
+	if err := k.maybeUpdateSubnetTags(); err != nil {
+		return err
+	}
+	return k.karpenterInstaller.Install(context.Background())
+}
+
+// makeNodeGroupStackName generates the name of the Karpenter stack identified by its name, isolated by the cluster this StackCollection operates on
+func (k *karpenterTask) makeKarpenterStackName() string {
+	return fmt.Sprintf("eksctl-%s-karpenter", k.cfg.Metadata.Name)
+}
+
+// maybeUpdateSubnetTags will check if the kubernetes.io/cluster tag is present on the subnets.
+// if not, it will create them.
+func (k *karpenterTask) maybeUpdateSubnetTags() error {
+	var ids []string
+	for _, subnet := range k.cfg.VPC.Subnets.Private {
+		ids = append(ids, subnet.ID)
+	}
+	for _, subnet := range k.cfg.VPC.Subnets.Public {
+		ids = append(ids, subnet.ID)
+	}
+	sort.Strings(ids)
+	input := &ec2.DescribeSubnetsInput{
+		SubnetIds: aws.StringSlice(ids),
+	}
+	output, err := k.ec2API.DescribeSubnets(input)
+	if err != nil {
+		return fmt.Errorf("failed to describe subnets: %w", err)
+	}
+
+	clusterTag := fmt.Sprintf(kubernetesTagFormat, k.cfg.Metadata.Name)
+
+	var updateSubnets []string
+	for _, subnet := range output.Subnets {
+		hasTag := false
+		for _, tag := range subnet.Tags {
+			if aws.StringValue(tag.Key) == clusterTag {
+				hasTag = true
+				break
+			}
+		}
+		if !hasTag {
+			updateSubnets = append(updateSubnets, *subnet.SubnetId)
+		}
+	}
```

Comment:
```suggestion
	var updateSubnets []string
	for _, subnet := range output.Subnets {
		for _, tag := range subnet.Tags {
			if aws.StringValue(tag.Key) == clusterTag {
				updateSubnets = append(updateSubnets, *subnet.SubnetId)
			}
		}
	}
```

keys are unique afaik so this should be okay

Reply:
The point here is to only update if the there is no such tag. We stop searching the tags once we find that it has the tag.

Reply:
oh sorry completely misread!